File: alcura_ipd_ext/alcura_ipd_extensions/report/live_bed_board/live_bed_board.js
// Copyright (c) 2026, Alcura and contributors
// For license information, please see license.txt

frappe.query_reports["Live Bed Board"] = {
	filters: [
		{
			fieldname: "ward",
			label: __("Ward"),
			fieldtype: "Link",
			options: "Hospital Ward",
			get_query: () => ({ filters: { is_active: 1 } }),
		},
		{
			fieldname: "room_type",
			label: __("Room Type"),
			fieldtype: "Link",
			options: "Healthcare Service Unit Type",
			get_query: () => ({ filters: { inpatient_occupancy: 1 } }),
		},
		{
			fieldname: "floor",
			label: __("Floor"),
			fieldtype: "Data",
		},
		{
			fieldname: "critical_care_only",
			label: __("Critical Care Only"),
			fieldtype: "Check",
			default: 0,
		},
		{
			fieldname: "gender",
			label: __("Gender"),
			fieldtype: "Select",
			options: "\nAll\nMale Only\nFemale Only",
			default: "",
		},
		{
			fieldname: "isolation_only",
			label: __("Isolation Only"),
			fieldtype: "Check",
			default: 0,
		},
		{
			fieldname: "payer_type",
			label: __("Payer Type"),
			fieldtype: "Select",
			options: "\nCash\nCorporate\nTPA",
			default: "",
			on_change: function () {
				const payer_type = frappe.query_report.get_filter_value("payer_type");
				const payer_filter = frappe.query_report.get_filter("payer");
				if (payer_type === "Corporate" || payer_type === "TPA") {
					payer_filter.df.hidden = 0;
					payer_filter.refresh();
				} else {
					frappe.query_report.set_filter_value("payer", "");
					payer_filter.df.hidden = 1;
					payer_filter.refresh();
				}
			},
		},
		{
			fieldname: "payer",
			label: __("Payer"),
			fieldtype: "Link",
			options: "Customer",
			hidden: 1,
			get_query: () => ({
				filters: { disabled: 0 },
			}),
		},
		{
			fieldname: "show_unavailable",
			label: __("Show Unavailable"),
			fieldtype: "Check",
			default: 0,
		},
	],

	formatter: function (value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);

		if (column.fieldname === "availability" && data) {
			const avail = data.availability;
			if (avail === "Available") {
				value = `<span class="indicator-pill green">${__("Available")}</span>`;
			} else if (avail === "Occupied") {
				value = `<span class="indicator-pill red">${__("Occupied")}</span>`;
			} else if (avail === "Maintenance") {
				value = `<span class="indicator-pill orange">${__("Maintenance")}</span>`;
			} else if (avail === "Infection Block") {
				value = `<span class="indicator-pill red">${__("Infection Block")}</span>`;
			} else if (avail === "Dirty") {
				value = `<span class="indicator-pill orange">${__("Dirty")}</span>`;
			} else if (avail === "Cleaning") {
				value = `<span class="indicator-pill yellow">${__("Cleaning")}</span>`;
			}
		}

		if (column.fieldname === "housekeeping_status" && data) {
			const hs = data.housekeeping_status;
			if (hs === "Clean") {
				value = `<span class="indicator-pill green">${__("Clean")}</span>`;
			} else if (hs === "Dirty") {
				value = `<span class="indicator-pill orange">${__("Dirty")}</span>`;
			} else if (hs === "In Progress") {
				value = `<span class="indicator-pill yellow">${__("In Progress")}</span>`;
			}
		}

		if (column.fieldname === "payer_eligible" && data) {
			if (data.payer_eligible === "Yes") {
				value = `<span class="indicator-pill green">${__("Yes")}</span>`;
			} else if (data.payer_eligible === "No") {
				value = `<span class="indicator-pill red">${__("No")}</span>`;
			}
		}

		return value;
	},

	onload: function (report) {
		report.page.add_inner_button(__("Refresh"), function () {
			report.refresh();
		});
	},
};

File: alcura_ipd_ext/alcura_ipd_extensions/report/live_bed_board/live_bed_board.py
"""Server-side entry point for the Live Bed Board Script Report."""

from __future__ import annotations

from alcura_ipd_ext.services.bed_availability_service import (
	get_available_beds,
	get_bed_board_summary,
)


def execute(filters: dict | None = None):
	filters = filters or {}
	columns = _get_columns(filters)
	data = get_available_beds(filters)
	summary = _get_report_summary(filters)

	return columns, data, None, None, summary


def _get_columns(filters: dict) -> list[dict]:
	"""Build column definitions. Payer columns appear only when a payer filter is active."""
	columns = [
		{
			"fieldname": "bed",
			"label": "Bed",
			"fieldtype": "Link",
			"options": "Hospital Bed",
			"width": 140,
		},
		{
			"fieldname": "bed_label",
			"label": "Label",
			"fieldtype": "Data",
			"width": 90,
		},
		{
			"fieldname": "room",
			"label": "Room",
			"fieldtype": "Link",
			"options": "Hospital Room",
			"width": 130,
		},
		{
			"fieldname": "ward",
			"label": "Ward",
			"fieldtype": "Link",
			"options": "Hospital Ward",
			"width": 140,
		},
		{
			"fieldname": "room_type",
			"label": "Room Type",
			"fieldtype": "Link",
			"options": "Healthcare Service Unit Type",
			"width": 140,
		},
		{
			"fieldname": "floor",
			"label": "Floor",
			"fieldtype": "Data",
			"width": 60,
		},
		{
			"fieldname": "availability",
			"label": "Availability",
			"fieldtype": "Data",
			"width": 110,
		},
		{
			"fieldname": "housekeeping_status",
			"label": "Housekeeping",
			"fieldtype": "Data",
			"width": 100,
		},
		{
			"fieldname": "gender_restriction",
			"label": "Gender",
			"fieldtype": "Data",
			"width": 90,
		},
		{
			"fieldname": "maintenance_hold",
			"label": "Maintenance",
			"fieldtype": "Check",
			"width": 50,
		},
		{
			"fieldname": "infection_block",
			"label": "Infection",
			"fieldtype": "Check",
			"width": 50,
		},
		{
			"fieldname": "equipment_notes",
			"label": "Equipment",
			"fieldtype": "Data",
			"width": 120,
		},
	]

	if filters.get("payer_type"):
		columns.extend([
			{
				"fieldname": "daily_rate",
				"label": "Daily Rate",
				"fieldtype": "Currency",
				"width": 100,
			},
			{
				"fieldname": "payer_eligible",
				"label": "Payer Eligible",
				"fieldtype": "Data",
				"width": 80,
			},
		])

	columns.extend([
		{
			"fieldname": "ward_classification",
			"label": "Ward Class",
			"fieldtype": "Data",
			"width": 100,
		},
		{
			"fieldname": "medical_department",
			"label": "Specialty",
			"fieldtype": "Link",
			"options": "Medical Department",
			"width": 120,
		},
	])

	return columns


def _get_report_summary(filters: dict) -> list[dict]:
	"""Build the report summary cards shown above the data table."""
	summary = get_bed_board_summary(filters)
	return [
		{
			"value": summary["total"],
			"indicator": "Blue",
			"label": "Total Beds",
			"datatype": "Int",
		},
		{
			"value": summary["available"],
			"indicator": "Green",
			"label": "Available",
			"datatype": "Int",
		},
		{
			"value": summary["occupied"],
			"indicator": "Red",
			"label": "Occupied",
			"datatype": "Int",
		},
		{
			"value": summary["blocked"],
			"indicator": "Orange",
			"label": "Blocked",
			"datatype": "Int",
		},
	]
